src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1793 @@
+    /// after the instruction has completed processing.
+    free_after_curr_inst: HashSet<PReg>,
+    /// The virtual registers of use operands that have been allocated in the current instruction
+    /// and for which edits had to be inserted to save and restore them because their constraint
+    /// doesn't allow the allocation they are expected to be in after the instruction.
+    /// 
+    /// This needs to be kept track of to generate the correct moves in the case where a
+    /// single virtual register is used multiple times in a single instruction with
+    /// different constraints.
+    use_vregs_saved_and_restored_in_curr_inst: HashSet<VReg>,
+    /// Physical registers that were used for late def operands and now free to be
+    /// reused for early operands in the current instruction.
+    ///
+    /// After late defs have been allocated, rather than returning their registers to
+    /// the free register list, it is added here to avoid the registers being used as
+    /// scratch registers.
+    /// 
+    /// For example, consider the following:
+    /// def v0, use v1
+    /// If the processing of v1 requires a stack-to-stack move, then a scratch register is
+    /// used and the instruction becomes:
+    /// def v0, use v1
+    /// move from stack0 to p0
+    /// move from p0 to stack1
+    /// 
+    /// Since scratch registers may be drawn from the free register list and v0 will be allocated and
+    /// deallocated before v1, then it's possible for the scratch register p0 to be v0's allocation,
+    /// which is incorrect because p0 will end up holding whatever is in stack0, not v0.
+    /// `freed_def_regs` avoids this by allowing the late def registers to be reused without making it
+    /// possible for this scratch register scenario to happen.
+    freed_def_pregs: PartedByRegClass<BTreeSet<PReg>>,
+    /// Used to keep track of which used vregs are seen for the first time
+    /// in the instruction, that is, if the vregs live past the current instruction.
+    /// This is used to determine whether or not reused operands
+    /// for reuse-input constraints should be restored after an instruction.
+    /// It's also used to determine if the an early operand can reuse a freed def operand's
+    /// allocation. And it's also used to determine the edits to be inserted when
+    /// allocating a use operand.
+    vregs_first_seen_in_curr_inst: HashSet<VReg>,
+    /// Used to keep track of which vregs have been allocated in the current instruction.
+    /// This is used to determine which edits to insert when allocating a use operand.
+    vregs_allocd_in_curr_inst: HashSet<VReg>,
+    /// All the safepoint instructions encountered during allocation and their blocks.
+    /// When allocation is completed, this contains all the safepoint instructions
+    /// in the function.
+    /// This is used to build the stackmap after allocation is complete.
+    safepoint_insts: Vec<(Block, Inst)>,
```

Comment:
Also as we noted in the call -- we can remove safepoint-related logic as we're likely to switch over to a non-regalloc-provided safepoint implementation in Cranelift soon. This should simplify things quite a bit!